src/builtin.jq

```diff
@@ -215,22 +215,9 @@ def fromstream(i): {x: null, e: false} as $init |
     else setpath(["e"]; $i[0]|length==1) end
   ; if .e then .x else empty end);
 def tostream:
-  {string:true,number:true,boolean:true,null:true} as $leaf_types |
-  . as $dot |
-  if $leaf_types[$dot|type] or length==0 then [[],$dot]
-  else
-    # We really need a _streaming_ form of `keys`.
-    # We can use `range` for arrays, but not for objects.
-    keys_unsorted as $keys |
-    $keys[-1] as $last|
-    ((# for each key
-      $keys[] | . as $key |
-      $dot[$key] | . as $dot |
-      # recurse on each key/value
-      tostream|.[0]|=[$key]+.),
-     # then add the closing marker
-     [[$last]])
-  end;
+  path(def r: (.[]?|r), .; r) as $p |
```

Comment:
I think you can make this just `path(..?) as $p |`.

Reply:
Recall that `..` is syntactic sugar for `recurse`, which is `def recurse: recurse(.[]?);`, and `recurse/1` is `def recurse(f): def r: ., (f | r); r;`.  However, inlining as here is best.

Reply:
.. (=recurse) is pre-order and we need post-order here to close nonempty objects/arrays.

Reply:
Ah, right.  Thanks for the clarification (and, I know, I missed this nugget of info for not reading the description together with the changes).